proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+         => a -> a -> Bool
+
+If it turns out that ``(==)`` is applied to a type for which no ``Eq`` instance exists, the error message becomes ::
+
+  * X does not implement equality.
+    Maybe you want to derive it using `deriving Eq`?
+  * In the expression: ...
+
+The second combinator is ``IfApartUnsafe`` (the naming shall become clear later). In short, ``IfApartUnsafe a b no yes`` rewrites to ``no`` if ``a`` and ``b`` are apart *at that point in the solving process* and to ``yes`` otherwise. Note that this is *extremely unsafe*, since any knowledge about ``a`` and ``b`` is lost if the second branch is taken, and for that reason we should only expose the variant ::
+
+    type IfApart a b no yes = IfApartUnsafe a b no (a ~ b, yes)
+
+Using this synonym we can give a better type to ``(==)`` ::
+
+    (==) :: IfApart a b
+              (TypeError (Text "== is applied to arguments of different types"))
+              (IfUndischarged (Eq a) (ShowType a :<>: ...))
+         => a -> b -> Bool
+
+One of the nice things about ``IfUndischarged`` and ``IfApartUnsafe`` being combinators is that we can abstract error patterns using type-level programming such as type families. See the following `prototype implementation <https://git.science.uu.nl/f100183/ghc/blob/wip/when-not/libraries/base/GHC/TypeErrors.hs>`_ of a proposed ``GHC.TypeErrors`` module. Using it we could write ::
+
+    (==) :: CustomErrors [ a :~/: b :=>: Text "== is applied to arguments of different types"
+                         , Undischarged (Eq a) :=>: ShowType a :<>: Text " does not implement equality."]
+         => a -> b -> Bool
+
+You might have noticed that I have emphasized "at that point in the solving process" when describing ``IfApartUnsafe``. The reason is that we cannot defer indefinitely every apartness check, so sometimes we need to take the ``yes`` branch and continue. Thus, the order in which constraints are solved becomes quite important. I propose to change the default from "don't care" to an order in which constraints coming from arguments to a function are given priority with respect to those coming from the application itself.
+
+Updates to the solver
+~~~~~~~~~~~~~~~~~~~~~
+Inside GHC solver each constraint would come with an optional attached message and a priority number. Whenever a new item has to be taken out of the work list, the ones with highest priority should come before. If more than one constraint have the same priority, then the choice is done as now -- first canonicalization, then interaction, and so on.
+
+Whenever the solver finds a ``IfUndischarged c msg``, it emits the constraint ``c`` with the additional message attached. This message has no influence on the rest of the solving process, but if it ends with errors and the constraint should be reported, the message is used. Indeed, we could see its definition as (note that ``UndecidableSuperClasses`` is required) ::
```

Comment:
This is the behavior for Wanteds. I assume Givens will just ignore the message. Or will the message be incorporated into the text GHC produces in the case of inaccessible code (which is what a Given becomes when it is insoluble)?